# Functions

In this tutorial, we'll learn how to use functions in Nain.

A function is a block of code that can be called from anywhere in the program.

The `main` function is the entry point of the program.

## 1.1 Declaring functions

To declare a function, you need to use the `fn` keyword.

```nain
fn isLightOn() {
    print!("The light is on!")
}

fn main() {
    isLightOn()
}
```

## 1.2 Calling functions

As you have seen on the example, you can call a function by using its name followed by parentheses.

```nain
fn isLightOn() {
    print!("The light is on!")
}

fn main() {
    isLightOn()
}
```

You might wonder why do we call the `isLightOn` function with parentheses.

This is because of **argument passing**.
If you don't know what argument passing is, don't worry, we'll explain it in the next chapter.

## 1.3 Arguments

When calling a function, you can pass arguments to it.

Arguments behave like variables, so you can use them as if they were numbers or strings.

The difference:
They are not declared with the `let` keyword, but in the parentheses.
They can only be used inside the function they are declared in.

Let's see an example:

```nain
fn addTwoNumbers(x: int, y: int) { // you can specify the data type of the argument
    return x + y
}
```

Now, we can call the `addTwoNumbers` function like this:

```nain
fn main() {
    let result = addTwoNumbers(10, 20)
    print!(result) // prints 30
}

```

As you can see, arguments are separated by a comma.

You might wonder what the `return` keyword does.

## 1.4 Returning values

It's used to return a value from a function.

So we can use a function's return value like a variable with:

```nain
fn main() {
    let result = addTwoNumbers(10, 20)
    print!(result) // prints 30
    print!(addTwoNumbers(10, 20)) // prints 30
}
```

## 1.5 Functions as arguments

Functions can be passed as arguments to other functions.

Let's see an example:

```nain
fn addTwoNumbers(x: int, y: int) {
    return x + y
}

fn main() {
    let result1 = addTwoNumbers(10, 20)
    let result2 = addTwoNumbers(addTwoNumbers(5,5), 20)
    print!(result) // prints 30
}
```

In the example above, we call the `addTwoNumbers` function twice.
The first time, we pass the arguments `10` and `20`.
The second time, we pass the result of the first call as the first argument, and `20` as the second argument.

The result of the second call is `35`.

## What's next?

In the next tutorial, we'll learn about loops and conditionals.
